### PR TITLE
Removed softmax Int16 dependency on std::vector

### PR DESCRIPTION
**Pull request by yair-ehrenwald**
This addresses the issue raised in https://github.com/tensorflow/tensorflow/issues/43126

The reference implementation of int16 softmax uses std::vector, this PR removes that dependency. 
The approach I took was rewriting the loop to in order to calculate the exponent twice instead of saving it after the first calculation - essentially paying with CPU time for memory. Slightly less efficient but embedded friendly, especially considering that most embedded platforms will just use it as reference and probably use an optimized implementation.

### CODE REVIEW

**Comment by advaitjain** [comment]
Thanks for the PR.

I'm ok with the change from the perspective of portability of the reference implementation for TF Micro, but adding in the folks involved in PR #35996 to get their opinion.

@psunn, @jdduke, @jianlijianli : thoughts on this PR?

@freddan80  and @patriklaurell as other folks who might want to weigh in.

**Review by freddan80 — COMMENTED** [review]
Looks good to me!

**Comment by freddan80** [comment]
Tagging @psunn who submitted the original implementation of int16 softmax

File: tensorflow/lite/kernels/internal/reference/softmax.h [hunk 1]
```diff
@@ -145,6 +144,20 @@ inline void Softmax(const SoftmaxParams& params,
   }
 }
 
+int16_t CalculateExp(const SoftmaxParams& params, const int16_t* input_data,
```
Comment (psunn) [inline tensorflow/lite/kernels/internal/reference/softmax.h:147]: `inline int16_t CalculateExp`
currently test failed with multiple definition

meanwhile, I think it's ok to remove container and calculate `exp_result_Q015` as needed.

Comment (yair-ehrenwald) [reply tensorflow/lite/kernels/internal/reference/softmax.h:147]: Which of the tests failed? Everything passed for me.
If the issue is doing the calculation in CalculateExpI can move it back into the main loop, moved it out for readability.

Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/softmax.h:147]: Hi, it appears compile error, when I run test:
`bazel test tensorflow/lite/kernels:activations_test --test_output=all`

Comment (yair-ehrenwald) [reply tensorflow/lite/kernels/internal/reference/softmax.h]: Ah, I see, multiple defs. I inlined the CalculateExp function and all seem to pass, see latest commit.

**Comment by Tessil** [comment]
Instead of calling `CalculateExp` two times for each `c` wouldn't it be possible to either:
* Pass a buffer of size `c` in parameter to store the intermediate results.
* Use the `output_data` as intermediate buffer to store the results of  `CalculateExp`. We'll only overwrite this result when we store the final output and thus when we don't need the result of `CalculateExp` anymore.

**Comment by psunn** [comment]
@Tessil for second thoughts

**Comment by yair-ehrenwald** [comment]
Yes, possible. I considered using the output as scratch as it is ideal for it, however I thought that might be too much of a dirty trick for a reference implementation. What do you think?

**Comment by Tessil** [comment]
Personally I think that replacing `std::vector<int16_t> exp_result_Q015(depth);` with `int16_t* exp_result_Q015 = output_data + i * depth;` in the original code with a little comment regarding the `output_data` re-usage and the overriding would be alright but it'd be best to wait what the others think of it.

**Comment by yair-ehrenwald** [comment]
> Personally I think that replacing `std::vector<int16_t> exp_result_Q015(depth);` with `int16_t* exp_result_Q015 = output_data + i * depth;` in the original code with a little comment regarding the `output_data` re-usage and the overriding would be alright but it'd be best to wait what the others think of it.

See commit from just a minute ago :)

File: tensorflow/lite/kernels/internal/reference/softmax.h [hunk 2]
```diff
@@ -145,6 +144,21 @@ inline void Softmax(const SoftmaxParams& params,
   }
 }
 
+inline int16_t SoftMaxCalculateExp(const SoftmaxParams& params,
+                                   const int16_t* input_data, const int depth,
+                                   int16_t max_in_row, int i, int c) {
+  int32_t input_diff = input_data[i * depth + c] - max_in_row;
+  // scale the input_diff such that [-65535, 0] correspond to [-10.0, 0.0]
```
Comment (advaitjain) [inline tensorflow/lite/kernels/internal/reference/softmax.h:151]: I know this is existing code so feel free to ignore, but this comment seems stale. Is there something special about the range [-10, 0] that I am missing?

**Review by advaitjain — CHANGES_REQUESTED** [review]
I'm ok with the optimization of reusing output_data.

My only request for changes are some documentation improvements and then we can merge this.

File: tensorflow/lite/kernels/internal/reference/softmax.h (hunk 2 above)
Comment (psunn) [reply tensorflow/lite/kernels/internal/reference/softmax.h:151]: I think it's better to keep it.

exp lut generated with range [-10, 0], as exp(-10) is insignificant to accumulate, such logic can be find in `SoftmaxPrepare` function

**Comment by sicong-li-arm** [comment]
> Thanks for the PR.
> 
> I'm ok with the change from the perspective of portability of the reference implementation for TF Micro, but adding in the folks involved in PR #35996 to get their opinion.
> 
> @psunn, @jdduke, @jianlijianli : thoughts on this PR?
> 
> @freddan80 and @sicong-li-arm as other folks who might want to weigh in.

Looks good to me as well! Thanks @yair-ehrenwald for taking this issue :)

File: tensorflow/lite/kernels/internal/reference/softmax.h [hunk 3]
```diff
@@ -165,27 +179,14 @@ inline void SoftmaxInt16(const SoftmaxParams& params,
     }
 
     // Compute exp(input - max_input)
```
[the remark this answers is not in the document]
Comment (yair-ehrenwald) [reply tensorflow/lite/kernels/internal/reference/softmax.h]: Done

File: tensorflow/lite/kernels/internal/reference/softmax.h [hunk 4]
```diff
@@ -165,27 +179,14 @@ inline void SoftmaxInt16(const SoftmaxParams& params,
...
-      int16_t sat_sym_scaled_diff =
-          std::min(std::max(sym_scaled_diff, static_cast<int32_t>(-32768)),
-                   static_cast<int32_t>(32767));
-      // apply the exp() LUT activation function
-      exp_result_Q015[c] =
-          generic_int16_table_lookup(sat_sym_scaled_diff, params.exp_lut);
-    }
-
     // sum_of_exps is a Q16.15 fixed point format.
+	//the output buffer is used as scratch memory and overwritten when the results are written.
```
[the remark this answers is not in the document]
Comment (yair-ehrenwald) [reply tensorflow/lite/kernels/internal/reference/softmax.h]: See if acceptable now..

File: tensorflow/lite/kernels/internal/reference/softmax.h (hunk 2 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/kernels/internal/reference/softmax.h:152]: I added a comment regarding @psunn 's explanation.

**Review by advaitjain — APPROVED** [review]
looks great, thanks for the quick fix!
